app/server-utils/url-to-iframe.ts

```diff
@@ -0,0 +1,35 @@
+/**
+ * Replaces all of the URLs in text with iframes of the URLs
+ * @param text text in which URLs should be replaced
+ */
+export const urlToIframe = (text: string): string => {
+  const whitelistedHosts: Record<string, HostConfig> = {
+    'aisafety.world': {sandboxValue: 'allow-scripts allow-same-origin'},
+  }
+
+  // Regex is from: https://stackoverflow.com/a/26764609
+  const anchorTagRegex = new RegExp(/<a[\s]+([^>]+)>((?:.(?!<\/a>))*.)<\/a>/gi)
```

Comment:
something like `<a rel="alternate" href="https://aisafety.world/" target="_blank">https://aisafety.world/</a>` will break this. A simple fix would be to change the regex to `<a\s+[^>]*?(href="[^>]+")[^>]*?>((?:.(?!<\/a>))*.)<\/a>`. That being said, regexing HTML is generally a bad idea. [This](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454) is the classic reference.

A more generic option would be to use DOMParser, e.g.:
```
   const parser = new DOMParser();
    const htmlDoc = parser.parseFromString(text, 'text/html');
    for(const a of htmlDoc.getElementsByTagName('a')) {
        const host = new URL(a.href).host
        const hostConfig = whitelistedHosts[host]
        if (hostConfig && a.href.trim() == a.innerHTML.trim()) {
            a.outerHTML = `<iframe src="${a.href}" sandbox="${hostConfig.sandboxValue}"></iframe>`
        }
    }
    return htmlDoc.body.innerHTML
```
The main problem with this, of course, is that it will only work in the browser. So you'd have to move the call from `stampy.ts` to `routes/questions/$question.tsx`. The thing with that is Remix calls routes both in the browser (which is what you want) and on the backend (which will cause an error). There are various ways to handle this, the simplest being just wrapping the parser like this:
```
    let parser
    try {
        parser = new DOMParser();
    } catch (error) {
        return text
    }
```

A nice thing about your solution is that it's cached, so the transformation is called once per question, per API call. Moving this to the frontend would break that, and parsing is generally a lot slower that a basic regex. That being said, unless the questions get too bulky, it shouldn't be a problem?

Reply:
That being said, regex on markdown would be fine. How about moving the regexes earlier, before the contents get transformed from markdown to HTML?

Reply:
> Remix calls routes both in the browser (which is what you want) and on the backend (which will cause an error)

Is this referring to serverside rendering from Remix?

Reply:
yes. If you call this from `stampy.ts::renderText()`, you'll be able to play with markdown, rather than html, which is a lot simpler

Reply:
@mruwnik I've been thinking about this a bit more and I'm not sure parsing the markdown is better than parsing the html. The main reason is that there is no mapping in markdown to an iframe (which is what we need to get to)
In other words:
- in html, the mapping is from `<a>` tag to `<iframe>`
- in markdown, the mapping would be from `[]()` links to ???

Maybe one could
1. find the relevant links in the markdown
2. replace with a special symbol
3. convert this special symbol to `iframe`

Or do you see more clearly how it would work in markdown?

I mean, I see this is merged already, so maybe not a big deal

Reply:
I'm guessing you could do pretty much exactly the same as you do here, but the regex would be simpler. Though I don't know if the HTML generator would correctly handle embedded HTML. Could be checked if you're feeling bored...